File: native/ios-app/App/Views/Chat/FireChatBusPayload.swift
import Foundation

extension Notification.Name {
    static let fireChatMessageBusEvent = Notification.Name("fire.chat.messageBusEvent")
}

enum FireChatBusPayload {
    static func event(
        from busEvent: MessageBusEventState,
        fallbackChannelID: UInt64?,
        baseURLString: String
    ) -> ChatBusEventState {
        guard let payload = busEvent.payloadJson else {
            return .ignored
        }
        return chatBusEventFromPayload(
            payloadJson: payload,
            eventType: busEvent.detailEventType ?? busEvent.messageType,
            fallbackChannelId: fallbackChannelID ?? busEvent.topicId,
            baseUrl: baseURLString
        )
    }

    static func chatMessage(
        from event: MessageBusEventState,
        fallbackChannelID: UInt64?,
        baseURLString: String
    ) -> ChatMessageState? {
        switch Self.event(from: event, fallbackChannelID: fallbackChannelID, baseURLString: baseURLString) {
        case let .messageUpsert(message):
            return message
        default:
            guard let payload = event.payloadJson else { return nil }
            return chatMessageFromBusPayload(
                payloadJson: payload,
                fallbackChannelId: fallbackChannelID ?? event.topicId,
                baseUrl: baseURLString
            )
        }
    }

    static func channel(
        from event: MessageBusEventState,
        baseURLString: String
    ) -> ChatChannelState? {
        switch Self.event(from: event, fallbackChannelID: event.topicId, baseURLString: baseURLString) {
        case let .channelUpsert(channel):
            return channel
        default:
            guard let payload = event.payloadJson else { return nil }
            return chatChannelFromBusPayload(payloadJson: payload, baseUrl: baseURLString)
        }
    }
}

File: native/android-app/src/main/java/com/fire/app/ui/chat/FireChatBusPayload.kt
package com.fire.app.ui.chat

import uniffi.fire_uniffi_chat.ChatBusEventState
import uniffi.fire_uniffi_chat.ChatChannelState
import uniffi.fire_uniffi_chat.ChatMessageState
import uniffi.fire_uniffi_chat.chatBusEventFromPayload
import uniffi.fire_uniffi_chat.chatChannelFromBusPayload
import uniffi.fire_uniffi_chat.chatMessageFromBusPayload
import uniffi.fire_uniffi_messagebus.MessageBusEventState

object FireChatBusPayload {
    fun event(
        busEvent: MessageBusEventState,
        fallbackChannelId: ULong?,
        baseUrl: String,
    ): ChatBusEventState {
        val payload = busEvent.payloadJson ?: return ChatBusEventState.Ignored
        return chatBusEventFromPayload(
            payloadJson = payload,
            eventType = busEvent.detailEventType ?: busEvent.messageType,
            fallbackChannelId = fallbackChannelId ?: busEvent.topicId,
            baseUrl = baseUrl,
        )
    }

    fun chatMessage(
        busEvent: MessageBusEventState,
        fallbackChannelId: ULong?,
        baseUrl: String,
    ): ChatMessageState? {
        return when (val parsed = event(busEvent, fallbackChannelId, baseUrl)) {
            is ChatBusEventState.MessageUpsert -> parsed.message
            else -> busEvent.payloadJson?.let { payload ->
                chatMessageFromBusPayload(
                    payloadJson = payload,
                    fallbackChannelId = fallbackChannelId ?: busEvent.topicId,
                    baseUrl = baseUrl,
                )
            }
        }
    }

    fun channel(
        busEvent: MessageBusEventState,
        baseUrl: String,
    ): ChatChannelState? {
        return when (val parsed = event(busEvent, busEvent.topicId, baseUrl)) {
            is ChatBusEventState.ChannelUpsert -> parsed.channel
            else -> busEvent.payloadJson?.let { payload ->
                chatChannelFromBusPayload(payloadJson = payload, baseUrl = baseUrl)
            }
        }
    }
}
